rust/src/array_data.rs

```diff
@@ -0,0 +1,264 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use bitmap::Bitmap;
+use buffer::Buffer;
+use datatypes::DataType;
+use util::bit_util::count_set_bits;
+
+#[derive(PartialEq, Debug)]
+pub struct ArrayData {
+    /// The data type for this array data
+    data_type: DataType,
+
+    /// The number of elements in this array data
+    length: i64,
+
+    /// The number of null elements in this array data
+    null_count: i64,
+
+    /// The offset into this array data.
+    offset: i64,
+
+    /// The buffers for this array data. Note that depending on the array types, this
+    /// could hold different types of buffers (e.g., value buffer, value offset buffer)
+    /// at different positions.
+    buffers: Vec<Buffer>,
+
+    /// The child(ren) of this array. Only non-empty for `ListArray` and `StructArray`.
+    child_data: Vec<ArrayDataRef>,
+
+    /// The null bit map. A `None` value for this indicates all values are non-null in
+    /// this array.
+    null_bitmap: Option<Bitmap>,
+}
+
+pub type ArrayDataRef = Arc<ArrayData>;
+pub const UNKNOWN_NULL_COUNT: i64 = -1;
+
+impl ArrayData {
+    pub fn new(
+        data_type: DataType,
+        length: i64,
+        mut null_count: i64,
+        null_bit_buffer: Option<Buffer>,
+        offset: i64,
+        buffers: Vec<Buffer>,
+        child_data: Vec<ArrayDataRef>,
+    ) -> Self {
+        if null_count < 0 {
+            null_count = if let Some(ref buf) = null_bit_buffer {
+                count_set_bits(buf.data())
```

Comment:
For the record, in the C++ implementation we call `count_set_bits` lazily (when a caller first asks for the null count). YMMV.

Reply:
Yes I thought about that but moving this to `null_count()` will require it to take a `&mut self` instead of `&self` (since now it needs to modify `self.null_count`), which is not very desirable.

Reply:
Apparently Rust allows interior mutability, which is adequate for such cached properties:
https://ricardomartins.cc/2016/06/08/interior-mutability

Reply:
Yes, I forgot about `Cell`. Will change.

Reply:
Actually @pitrou : the `Cell` is not thread-safe in this case so we cannot use it. I thought about using `AtomicI64` but it is nightly-only. I'm not sure how significant this is for performance reason - can we leave it as a TODO for now?